File: util/src/main/scala/com/flink/common/source/KafkaSourceManager.scala
package com.flink.common.source

import com.flink.common.deserialize.TopicMessageDeserialize
import com.flink.common.kafka.KafkaManager
import com.flink.common.kafka.KafkaManager.KafkaMessge
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer010
import scala.collection.JavaConverters._
object KafkaSourceManager {

  /**
    *
    * @param topic
    * @param broker
    * @param reset
    * @param groupid
    * @param setStartFromTimestamp
    * @return
    */
  def getKafkaSource(
      topic: String,
      broker: String,
      reset: String,
      groupid: String,
      setStartFromTimestamp: Long = 0L): FlinkKafkaConsumer010[KafkaMessge] = {
    val kafkasource = new FlinkKafkaConsumer010[KafkaMessge](
      topic.split(",").toList.asJava,
      new TopicMessageDeserialize(),
      KafkaManager.getKafkaParam(broker, groupid))
    kafkasource.setCommitOffsetsOnCheckpoints(true)
    if (setStartFromTimestamp > 0) {
      kafkasource.setStartFromTimestamp(setStartFromTimestamp)
    } else {
      if (reset == "latest") {
        kafkasource.setStartFromLatest()
      } else if (reset == "earliest") {
        kafkasource.setStartFromEarliest()
      }
    }
    kafkasource
  }

}

File: flink-util/src/main/java/com/flink/common/java/manager/KafkaSourceManager.java
package com.flink.common.java.manager;

import com.flink.common.deserialize.TopicOffsetJsonEventtimeDeserialize;
import com.flink.common.deserialize.TopicOffsetTimeStampMsgDeserialize;
import com.flink.common.kafka.KafkaManager;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer010;
import org.apache.flink.streaming.connectors.kafka.KafkaDeserializationSchema;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;

public class KafkaSourceManager {

    private static <T> FlinkKafkaConsumer010<T> getKafkaSource(
            String topic,
            String broker,
            String reset,
            KafkaDeserializationSchema<T> kds) {
        FlinkKafkaConsumer010<T> kafkasource = KafkaManager.getKafkaSource(
                topic,
                broker,
                kds);
        kafkasource.setCommitOffsetsOnCheckpoints(true);
        if (reset == "earliest") {
            kafkasource.setStartFromEarliest(); //不加这个默认是从上次消费
        } else if (reset == "latest") {
            kafkasource.setStartFromLatest();
        }
        return kafkasource;
    }


    public static <T> DataStreamSource getKafkaDataStream(StreamExecutionEnvironment streamEnv,
                                                          String topic,
                                                          String broker,
                                                          String reset,
                                                          KafkaDeserializationSchema<T> kds) {
        return streamEnv.addSource(getKafkaSource(topic, broker, reset, kds));
    }

    public static Table getStreamTable(StreamTableEnvironment streamTableEnv, DataStreamSource source ,String fields) {
        return streamTableEnv.fromDataStream(source, fields);
    }


}
